File: frontend/src/utils/permission.ts
import { useAuthStore } from '@/store/auth'

/**
 * 检查用户是否有指定权限
 */
export const hasPermission = (resource: string, action: string): boolean => {
  const { user } = useAuthStore.getState()
  if (!user || !user.roles) {
    return false
  }

  // 检查用户的所有角色权限
  for (const role of user.roles) {
    if (role.permissions) {
      for (const permission of role.permissions) {
        if (permission.resource === resource && permission.action === action) {
          return true
        }
      }
    }
  }

  return false
}

/**
 * 检查用户是否有指定资源权限（任意操作）
 */
export const hasResourcePermission = (resource: string): boolean => {
  const { user } = useAuthStore.getState()
  if (!user || !user.roles) {
    return false
  }

  for (const role of user.roles) {
    if (role.permissions) {
      for (const permission of role.permissions) {
        if (permission.resource === resource) {
          return true
        }
      }
    }
  }

  return false
}

File: backend/internal/model/permission.go
package model

// Permission 权限模型
type Permission struct {
	BaseModel
	Name        string `gorm:"type:varchar(50);not null" json:"name"`
	API         string `gorm:"type:varchar(255);not null;uniqueIndex:idx_api_method;comment:接口路径 如:/api/v1/users" json:"api"`
	Method      string `gorm:"type:varchar(10);not null;uniqueIndex:idx_api_method;comment:请求方法 如:GET,POST,PUT,DELETE" json:"method"`
	Resource    string `gorm:"type:varchar(50);not null;comment:资源名称 如:user,role" json:"resource"`
	Description string `gorm:"type:varchar(255)" json:"description"`
	Roles       []Role `gorm:"many2many:role_permissions;" json:"roles,omitempty"`
}

// TableName 指定表名
func (Permission) TableName() string {
	return "permissions"
}
